src/tempo_config.py

```diff
@@ -24,143 +22,184 @@ class InvalidConfigurationError(Exception):
+    """Memberlist schema."""
 
-    ingestion_rate: int = 0
-    ingestion_burst_size: int = 0
-    max_global_series_per_user: int = 0
-    ruler_max_rules_per_rule_group: int = 0
-    ruler_max_rule_groups_per_tenant: int = 0
+    abort_if_cluster_join_fails: bool
+    bind_port: int
+    join_members: List[str]
+    tls_enabled: Optional[bool] = False
+    tls_cert_path: Optional[str] = None
+    tls_key_path: Optional[str] = None
+    tls_ca_path: Optional[str] = None
+    tls_server_name: Optional[str] = None
 
 
-class Kvstore(BaseModel):
-    """Kvstore schema."""
+class ClientTLS(BaseModel):
+    """Client tls config schema."""
 
-    store: str = "memberlist"
+    tls_enabled: bool
+    tls_cert_path: str
+    tls_key_path: str
+    tls_ca_path: str
+    tls_server_name: str
 
 
-class Ring(BaseModel):
-    """Ring schema."""
+class Client(BaseModel):
+    """Client schema."""
 
-    kvstore: Kvstore
-    replication_factor: int = 3
+    grpc_client_config: ClientTLS
 
 
 class Distributor(BaseModel):
     """Distributor schema."""
 
-    ring: Ring
+    ring: Optional[Ring] = None
+    receivers: Dict[str, Any]
 
 
 class Ingester(BaseModel):
     """Ingester schema."""
 
-    ring: Ring
+    trace_idle_period: str
+    max_block_bytes: int
+    max_block_duration: str
+    lifecycler: Optional[Ring] = None
+
+
+class FrontendWorker(BaseModel):
+    """FrontendWorker schema."""
+
+    frontend_address: str
+    grpc_client_config: Optional[ClientTLS] = None
 
 
-class Ruler(BaseModel):
-    """Ruler schema."""
+class Querier(BaseModel):
+    """Querier schema."""
 
-    rule_path: str = "/data/ruler"
-    alertmanager_url: Optional[str]
+    frontend_worker: FrontendWorker
 
 
-class Alertmanager(BaseModel):
-    """Alertmanager schema."""
+class TLS(BaseModel):
+    """TLS configuration schema."""
 
-    data_dir: str = "/data/alertmanager"
-    external_url: Optional[str]
+    cert_file: str
+    key_file: str
+    client_ca_file: str
+    client_auth_type: Optional[str] = "VerifyClientCertIfGiven"
 
 
 class Server(BaseModel):
     """Server schema."""
 
-    http_tls_config: Dict[str, Dict[str, str]]
-    grpc_tls_config: Dict[str, Dict[str, str]]
+    http_listen_port: int
+    grpc_listen_port: int
+    http_tls_config: Optional[TLS] = None
+    grpc_tls_config: Optional[TLS] = None
+
+
+class Compaction(BaseModel):
+    """Compaction schema."""
+
+    compaction_window: str
+    max_compaction_objects: int
+    block_retention: str
+    compacted_block_retention: str
+    v2_out_buffer_bytes: int
+
+
+class Compactor(BaseModel):
+    """Compactor schema."""
 
+    ring: Optional[Ring] = None
+    compaction: Compaction
 
-class _S3ConfigData(BaseModel):
-    model_config = {"populate_by_name": True}
-    access_key_id: str = Field(alias="access-key")
+
+class Pool(BaseModel):
+    """Pool schema."""
+
+    max_workers: int
+    queue_depth: int
+
+
+class Wal(BaseModel):
+    """Wal schema."""
+
+    path: str
```

Comment:
I think pydantic takes `pathlib.Path` as well? perhaps that gives better validation?